app/src/main/java/me/tatarka/bindingcollectionadapter/sample/ImmutableViewModel.kt

```diff
@@ -21,56 +17,91 @@ class ImmutableViewModel : ViewModel(), ImmutableListeners {
-                    ) {
-                        val list =
-                            (0 until params.loadSize).map {
-                                ImmutableItem(
-                                    index = it + params.startPosition,
-                                    checked = false
-                                )
+
+                            override fun loadRange(
+                                    params: LoadRangeParams,
+                                    callback: LoadRangeCallback<Any>
+                            ) {
+                                val list =
+                                        (0 until params.loadSize).map {
+                                            ImmutableItem(
+                                                    index = it + params.startPosition,
+                                                    checked = false
+                                            )
+                                        }
+                                // Pretend we are slow
+                                Thread.sleep(1000)
+                                callback.onResult(list)
                             }
-                        // Pretend we are slow
-                        Thread.sleep(1000)
-                        callback.onResult(list)
-                    }
-                }
-        }, 20).build()
+                        }
+            }, PAGE_SIZE).build()
+
+    val pagedListV3: LiveData<PagingData<Any>> = Pager<Int, Any>(PagingConfig(
+            pageSize = PAGE_SIZE,
+            maxSize = 100,
+            prefetchDistance = PAGE_SIZE * 2,
+            enablePlaceholders = false)) {
+        object : PagingSource<Int, Any>() {
+            override suspend fun load(params: LoadParams<Int>): LoadResult<Int, Any> {
+                val safeKey = params.key ?: 0
+                val list =
+                        (0 until params.loadSize).map {
+                            ImmutableItem(
+                                    index = it + safeKey,
+                                    checked = false
+                            )
+                        }
+                // Pretend we are slow
+                delay(1000)
+
+                return LoadResult.Page(
+                        data = list,
+                        prevKey = if (safeKey == 0) null else (safeKey - params.loadSize),
+                        nextKey = if (safeKey >= TOTAL_COUNT - params.loadSize) null else (safeKey + params.loadSize),
+                        itemsBefore = safeKey,
+                        itemsAfter = TOTAL_COUNT - params.loadSize - safeKey
+                )
+            }
+        }
+    }.flow.asLiveData()
 
     val items = itemBindingOf<Any>(BR.item, R.layout.item_immutable)
 
+    val loadStateItems = OnItemBindLoadState.Builder<Any>()
+            .item(BR.item, R.layout.item_immutable)
+            .headerLoadState(BR.load, R.layout.network_state_item)
```

Comment:
Can we split this into its own Class based OnItemBind?  so we can separate the layouts for error/progress cases

Reply:
```
OnItemBindClass<LoadState>().apply {
            map<LoadState.Error>(BR.item, R.layout.network_state_item_error)
            map<LoadState.Loading>(ItemBinding.VAR_NONE, R.layout.network_state_item_progress)
        }.toItemBinding()
```

Reply:
I guess it would be pretty easy

```
.headerLoadState( OnItemBindClass<LoadState>().apply {
            map<LoadState.Error>(BR.item, R.layout.network_state_item_error)
            map<LoadState.Loading>(ItemBinding.VAR_NONE, R.layout.network_state_item_progress)
        } )
```